File: java/java-sdk/src/main/java/io/kestra/sdk/query/Query.java
/*
 * Kestra EE
 * All API operations, except for Superadmin-only endpoints, require a tenant identifier in the HTTP path.<br/> Endpoints designated as Superadmin-only are not tenant-scoped.
 *
 *
 *
 */


package io.kestra.sdk.query;

import io.kestra.sdk.model.QueryFilter;
import io.kestra.sdk.model.QueryFilterField;
import io.kestra.sdk.model.QueryFilterLogical;
import io.kestra.sdk.model.QueryFilterOp;

import java.util.ArrayList;
import java.util.List;

/**
 * Fluent constructor layer over {@link QueryFilter} for building complex AND/OR (one level of
 * nesting) query filters (issue #246). The result of {@link #where(QueryFilter)} is a plain
 * {@code List<QueryFilter>} that drops straight into every generated {@code *ByQuery} method.
 *
 * <pre>{@code
 * import static io.kestra.sdk.query.Query.*;
 *
 * List<QueryFilter> filters = where(and(
 *     eq(QueryFilterField.NAMESPACE, "prod"),
 *     or(
 *         eq(QueryFilterField.STATE, "RUNNING"),
 *         eq(QueryFilterField.STATE, "SUCCESS")
 *     )
 * ));
 * }</pre>
 */
public final class Query {

  private Query() {
  }

  /**
   * Builds a leaf filter node.
   */
  public static QueryFilter filter(QueryFilterField field, QueryFilterOp operation, Object value) {
    return new QueryFilter().field(field).operation(operation).value(value);
  }

  public static QueryFilter eq(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.EQUALS, value);
  }

  public static QueryFilter notEq(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.NOT_EQUALS, value);
  }

  public static QueryFilter in(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.IN, value);
  }

  public static QueryFilter notIn(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.NOT_IN, value);
  }

  public static QueryFilter contains(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.CONTAINS, value);
  }

  public static QueryFilter startsWith(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.STARTS_WITH, value);
  }

  public static QueryFilter endsWith(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.ENDS_WITH, value);
  }

  public static QueryFilter regex(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.REGEX, value);
  }

  public static QueryFilter prefix(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.PREFIX, value);
  }

  public static QueryFilter greaterThan(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.GREATER_THAN, value);
  }

  public static QueryFilter greaterThanOrEqual(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.GREATER_THAN_OR_EQUAL_TO, value);
  }

  public static QueryFilter lessThan(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.LESS_THAN, value);
  }

  public static QueryFilter lessThanOrEqual(QueryFilterField field, Object value) {
    return filter(field, QueryFilterOp.LESS_THAN_OR_EQUAL_TO, value);
  }

  /**
   * Builds an AND group. Null and empty-group children are dropped; an empty group returns
   * {@code null}; a single effective child is returned directly (flattened).
   */
  public static QueryFilter and(QueryFilter... children) {
    return group(QueryFilterLogical.AND, children);
  }

  /**
   * Builds an OR group. Same dropping/flattening rules as {@link #and(QueryFilter...)}.
   */
  public static QueryFilter or(QueryFilter... children) {
    return group(QueryFilterLogical.OR, children);
  }

  private static QueryFilter group(QueryFilterLogical logical, QueryFilter... children) {
    List<QueryFilter> effective = new ArrayList<QueryFilter>();
    if (children != null) {
      for (QueryFilter child : children) {
        if (child == null || isEmptyGroup(child)) {
          continue;
        }
        effective.add(child);
      }
    }
    if (effective.isEmpty()) {
      return null;
    }
    if (effective.size() == 1) {
      return effective.get(0);
    }
    return new QueryFilter().logical(logical).children(effective);
  }

  private static boolean isEmptyGroup(QueryFilter f) {
    return f.getField() == null && (f.getChildren() == null || f.getChildren().isEmpty());
  }

  /**
   * Turns a filter tree root into the SDK's list type accepted by {@code *ByQuery} methods.
   * A {@code null} root yields an empty list; a top-level AND group is flattened one level;
   * anything else yields a single-element list.
   */
  public static List<QueryFilter> where(QueryFilter root) {
    if (root == null) {
      return new ArrayList<QueryFilter>();
    }
    if (root.getLogical() == QueryFilterLogical.AND && root.getChildren() != null) {
      return new ArrayList<QueryFilter>(root.getChildren());
    }
    List<QueryFilter> result = new ArrayList<QueryFilter>();
    result.add(root);
    return result;
  }
}

File: python/python-sdk/kestrapy/query.py
"""Fluent DSL for building complex AND/OR query filters (issue #246).

Typed against the field/op/logical enums, this layer constructs
``QueryFilter`` trees that drop straight into any ``*ByQuery`` method::

    from kestrapy.query import where, and_, or_, eq, in_
    from kestrapy import QueryFilterField as F
    from kestrapy.models.state_type import StateType

    filters = where(
        and_(
            eq(F.NAMESPACE, "prod"),
            or_(
                eq(F.FLOW_ID, "a"),
                eq(F.FLOW_ID, "b"),
            ),
        )
    )
    client.search_executions(tenant="main", filters=filters)

``where`` returns ``list[QueryFilter]`` — the exact type every ``*ByQuery``
method already accepts.
"""

from typing import Any, List, Optional

from kestrapy.models.query_filter import QueryFilter
from kestrapy.models.query_filter_field import QueryFilterField
from kestrapy.models.query_filter_logical import QueryFilterLogical
from kestrapy.models.query_filter_op import QueryFilterOp

__all__ = [
    "where",
    "and_",
    "or_",
    "filter_",
    "eq",
    "not_eq",
    "in_",
    "not_in",
    "contains",
    "starts_with",
    "ends_with",
    "regex",
    "prefix",
    "gt",
    "gte",
    "lt",
    "lte",
]


def filter_(
    field: QueryFilterField, op: QueryFilterOp, value: Any = None
) -> QueryFilter:
    """Build a single leaf filter."""
    return QueryFilter(var_field=field, operation=op, value=value)


# --- convenience leaves -----------------------------------------------------

def eq(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.EQUALS, value)


def not_eq(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.NOT_EQUALS, value)


def in_(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.IN, value)


def not_in(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.NOT_IN, value)


def contains(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.CONTAINS, value)


def starts_with(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.STARTS_WITH, value)


def ends_with(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.ENDS_WITH, value)


def regex(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.REGEX, value)


def prefix(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.PREFIX, value)


def gt(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.GREATER_THAN, value)


def gte(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.GREATER_THAN_OR_EQUAL_TO, value)


def lt(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.LESS_THAN, value)


def lte(field: QueryFilterField, value: Any) -> QueryFilter:
    return filter_(field, QueryFilterOp.LESS_THAN_OR_EQUAL_TO, value)


# --- groups -----------------------------------------------------------------

def _is_empty_group(node: QueryFilter) -> bool:
    # Unified with the Java DSL's isEmptyGroup (issue #246 review): a node is an
    # empty group iff it is not a leaf (no field) and has no children. A leaf
    # (field set) is never an empty group.
    return (
        getattr(node, "var_field", None) is None
        and not getattr(node, "children", None)
    )


def _group(
    logical: QueryFilterLogical, children: tuple
) -> Optional[QueryFilter]:
    effective = [
        c for c in children if c is not None and not _is_empty_group(c)
    ]
    if not effective:
        return None
    if len(effective) == 1:
        return effective[0]
    return QueryFilter(logical=logical, children=effective)


def and_(*children: Optional[QueryFilter]) -> Optional[QueryFilter]:
    """AND-group its children (dropping None/empty; single-child flattens)."""
    return _group(QueryFilterLogical.AND, children)


def or_(*children: Optional[QueryFilter]) -> Optional[QueryFilter]:
    """OR-group its children (dropping None/empty; single-child flattens)."""
    return _group(QueryFilterLogical.OR, children)


# --- root -------------------------------------------------------------------

def where(root: Optional[QueryFilter]) -> List[QueryFilter]:
    """Turn a filter tree into the ``list[QueryFilter]`` ``*ByQuery`` expects.

    - ``None`` root -> ``[]``.
    - A top-level AND group -> its children flattened one level.
    - Anything else -> ``[root]``.
    """
    if root is None:
        return []
    logical = getattr(root, "logical", None)
    children = getattr(root, "children", None)
    if (
        logical is not None
        and str(logical.value if hasattr(logical, "value") else logical).lower()
        == "and"
        and children is not None
    ):
        return list(children)
    return [root]
